File: Ruby/BucketSort/stack_pointer_bucket_sort.rb
require 'benchmark'
require 'objspace'

class Rating
  attr_accessor :user_id, :movie_id, :rating, :timestamp
  def initialize(user_id, movie_id, rating, timestamp)
    @user_id = user_id
    @movie_id = movie_id
    @rating = rating
    @timestamp = timestamp
  end
  def to_s
    "#{@user_id},#{@movie_id},#{@rating},#{@timestamp}"
  end
end

class Node
  attr_accessor :rating, :next, :prev
  def initialize(rating)
    @rating = rating
    @next = nil
    @prev = nil
  end
end

class StackPointerBucketSort
  attr_accessor :top, :size
  def initialize
    @top = nil
    @size = 0
  end

  def push(rating)
    node = Node.new(rating)
    if @top.nil?
      @top = node
    else
      node.prev = @top
      @top.next = node
      @top = node
    end
    @size += 1
  end

  def bucket_sort
    base = @top
    base = base.prev while base&.prev

    buckets = Array.new(11) { [] }
    p = base
    while p
      idx = [(p.rating.rating * 2).floor, 10].min
      buckets[idx] << p
      p = p.next
    end

    # Reorganiza os nós
    new_top = nil
    prev = nil
    buckets.each do |bucket|
      bucket.each do |node|
        node.prev = prev
        node.next = nil
        prev.next = node if prev
        prev = node
        new_top = node # último da nova lista
      end
    end

    @top = new_top
  end
end

# --- Execução principal ---

if ARGV.size < 2
  puts "Uso: ruby stack_pointer_bucket_sort.rb <arquivo.csv> <N>"
  exit
end

path = ARGV[0]
n = ARGV[1].to_i
pilha = StackPointerBucketSort.new

File.open(path) do |f|
  f.readline
  f.each_line.take(n).each do |line|
    user_id, movie_id, rating, timestamp = line.strip.split(',')
    pilha.push(Rating.new(user_id.to_i, movie_id.to_i, rating.to_f, timestamp.to_i))
  end
end

mem_before = ObjectSpace.memsize_of_all
time = Benchmark.realtime { pilha.bucket_sort }
mem_after = ObjectSpace.memsize_of_all

puts "Ordenado #{n} registros em #{(time * 1000).round(2)} ms"
puts "Memória usada: #{((mem_after - mem_before) / 1024.0).round(2)} KB"

puts "Primeiros 100 registros ordenados:"
cur = pilha.top
cur = cur.prev while cur&.prev
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.next
end

puts "Últimos 100 registros ordenados:"
cur = pilha.top
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.prev
end

puts "Tamanho da pilha: #{pilha.size}"

File: JAVA/Bucket Sort/StackPointerBucketSort.java
import java.io.*;
import java.util.ArrayList;

public class StackPointerBucketSort
{
    static class Rating
    {
        int u,m;float r;long t;
        Rating(int u,int m,float r,long t){this.u=u;this.m=m;this.r=r;this.t=t;}
        public String toString(){return u+","+m+","+r+","+t;}
    }

    static class Node{Rating v;Node nxt,prv;Node(Rating v){this.v=v;}}

    Node top=null; int size=0;
    void push(Rating v){Node n=new Node(v);if(top!=null){n.prv=top;top.nxt=n;}top=n;size++;}

    void bucketSort()
    {
        Node base=top; while(base!=null&&base.prv!=null) base=base.prv;

        @SuppressWarnings("unchecked") ArrayList<Node>[] buckets=new ArrayList[11];
        for(int i=0;i<11;i++) buckets[i]=new ArrayList<>();
        for(Node p=base;p!=null;p=p.nxt) buckets[(int)Math.floor(p.v.r*2)].add(p);

        top=null; size=0;

        Node prev=null;
        for(ArrayList<Node>b:buckets)for(Node n:b)
        {
            n.nxt=null; n.prv=prev;
            if(prev!=null) prev.nxt=n;
            prev=n; if(top==null) top=n; size++;
        }

        top=prev;
    }

    public static void main(String[]a)
    throws Exception
    {
        if(a.length<2)
        {
            System.out.println("Uso: java StackPointerBucketSort <ratings.csv> <N>");
            return;
        }

        int N=Integer.parseInt(a[1]); StackPointerBucketSort S=new StackPointerBucketSort();

        try(BufferedReader br=new BufferedReader(new FileReader(a[0])))
        {
            String line=br.readLine();int c=0;
            while((line=br.readLine())!=null&&c<N)
            {
                String[] f=line.split(",");
                S.push(new Rating(Integer.parseInt(f[0]),Integer.parseInt(f[1]),
                                  Float.parseFloat(f[2]),Long.parseLong(f[3])));c++;
            }
        }

        long t0=System.nanoTime();S.bucketSort();long t1=System.nanoTime();
        System.out.printf("Bucket sort pilha ponteiro %d regs: %.2f ms%n",N,(t1-t0)/1e6);

        Node b=S.top; while(b!=null&&b.prv!=null) b=b.prv;
        for(int i=0;i<20&&b!=null;i++,b=b.nxt)System.out.println(b.v);
    }
}
